examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,213 @@
+        std_y_train = tf.cast(std_y_train, hps.dtype)
+        self.log_likelihood = log_mean_exp(log_likelihood, 0) / n_batch - \
+            tf.log(std_y_train)
+        y_pred_mean = tf.reduce_mean(y_inferred.distribution.mean, axis=0)
+        self.pred_mse = tf.reduce_mean((y_pred_mean - y_ph) ** 2)
+        self.pred_mse *= std_y_train ** 2
+
+    @zs.reuse('model')
+    def build_model(self, observed, n_particles, full_cov=False):
+        '''
+        Build the SVGP model.
+        Note that for inference, we only need the diagonal part of Cov[Y], as
+        ELBO equals sum over individual observations.
+        For visualization etc we may want a full covariance. Thus the argument
+        `full_cov`.
+        '''
+        hps = self._hps
+        with zs.BayesianNet(observed) as model:
+            Kzz_chol = tf.cholesky(self._kernel(self._z_pos, self._z_pos))
+            fZ = zs.MultivariateNormalCholesky(
+                'fZ', tf.zeros([hps.n_z], dtype=hps.dtype), Kzz_chol,
+                n_samples=n_particles)
+            # f(X)|f(Z) follows GP(0, K) gp_conditional
+            fx_given_fz = gp_conditional(
+                self._z_pos, fZ, self._x_ph, full_cov, self._kernel, 'fX',
+                Kzz_chol)
+            # Y|f(X) ~ N(f(X), noise_level * I)
+            noise_level = tf.get_variable(
+                'noise_level', shape=[], dtype=hps.dtype,
+                initializer=tf.constant_initializer(0.05))
+            noise_level = tf.nn.softplus(noise_level)
+            y = zs.Normal(
+                'y', mean=fx_given_fz, std=noise_level, group_ndims=1)
+        return model, y
+
+    def build_variational(self, n_particles):
+        hps = self._hps
+        with zs.BayesianNet() as variational:
+            z_mean = tf.get_variable(
+                'z/mean', [hps.n_z], hps.dtype, tf.zeros_initializer())
+            z_cov_raw = tf.get_variable(
+                'z/cov_raw', initializer=tf.eye(hps.n_z, dtype=hps.dtype))
+            z_cov_tril = tf.matrix_set_diag(
+                tf.matrix_band_part(z_cov_raw, -1, 0),
+                tf.nn.softplus(tf.matrix_diag_part(z_cov_raw)))
+            fZ = zs.MultivariateNormalCholesky(
+                'fZ', z_mean, z_cov_tril, n_samples=n_particles)
+            fx_given_fz = gp_conditional(
+                self._z_pos, fZ, self._x_ph, False, self._kernel, 'fX')
+        return variational
+
+    def infer_step(self, sess, x_batch, y_batch):
+        fd = {
+            self._x_ph: x_batch,
+            self._y_ph: y_batch,
+            self._n_particles_ph: self._hps.n_particles
+        }
+        return sess.run([self._infer_op, self._elbo], fd)[1]
+
+    def predict_step(self, sess, x_batch, y_batch):
+        fd = {
+            self._x_ph: x_batch,
+            self._y_ph: y_batch,
+            self._n_particles_ph: self._hps.n_particles_test
+        }
+        return sess.run([self.log_likelihood, self.pred_mse], fd)
+
```

Comment:
I still like the non-class version, which is more readable and fits our common paradigm. Any reason for the change?

Reply:
There will be too many arguments to pass around otherwise (x, z_pos, kernel, hps, etc). To me it makes sense to wrap these things to a context class, but I don't have strong preference. Should we revert it?

Reply:
due to the above discussion. consider reverting it now.